File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = n
    graph = [[] for _ in range(n+1)]
    
    for wire in wires:
        graph[wire[0]].append(wire[1])
        graph[wire[1]].append(wire[0])
        
    def bfs(s):
        visited = [False]*(n+1)
        cnt = 0
        q = deque()
        q.append(s)
        
        while q:
            num = q.popleft()
            visited[num] = True
            
            for i in graph[num]:
                if not visited[i]:
                    q.append(i)
                    cnt += 1
        
        return cnt
    
    for wire in wires:
        graph[wire[0]].remove(wire[1])
        graph[wire[1]].remove(wire[0])
        
        answer = min(abs(bfs(wire[0]) - bfs(wire[1])), answer)
        
        graph[wire[0]].append(wire[1])
        graph[wire[1]].append(wire[0])
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static int[][] maps;
    
    static int bfs(int n){
        boolean[] visited = new boolean[maps.length];
        int count = 0;
        Deque<Integer> queue = new ArrayDeque<>();
        
        visited[n] = true;
        queue.offerLast(n);
        
        while (!queue.isEmpty()){
            int cur = queue.pollFirst();
            
            for (int i = 0; i < maps.length; i++){
                if (maps[cur][i] == 1 && !visited[i]){
                    count += 1;
                    visited[i] = true;
                    queue.offerLast(i);
                }
            }
        }
        
        return count;
    }
    
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        maps = new int[n + 1][n + 1];
        
        for (int[] wire : wires){
            maps[wire[0]][wire[1]] = 1;
            maps[wire[1]][wire[0]] = 1;
        }
        
        for (int[] wire : wires){
            maps[wire[0]][wire[1]] = 0;
            maps[wire[1]][wire[0]] = 0;
            
            int wireA = bfs(wire[0]);
            int wireB = bfs(wire[1]);
            
            answer = Math.min(answer, Math.abs(wireA - wireB));
            
            maps[wire[0]][wire[1]] = 1;
            maps[wire[1]][wire[0]] = 1;
        }
        
        return answer;
    }
}